Mathlib/RingTheory/MvPowerSeries/Substitution.lean

```diff
@@ -0,0 +1,1096 @@
+    {A : Type*} [CommSemiring A] (R : Type*) [CommSemiring R] [Algebra A R]
+    (e : σ →₀ ℕ) (a : σ → A)  :
+    e.prod (fun s n ↦ ((a s • X s : MvPowerSeries σ R) ^ n))
+      = (e.prod fun s n ↦ (a s) ^ n) • monomial R e 1 := by
+  rw [Finsupp.prod_congr
+    (g2 := fun s n ↦ ((C σ R (algebraMap A R (a s)) * (X s : MvPowerSeries σ R)) ^ n))]
+  · have (a : A) (f : MvPowerSeries σ R) : a • f = (C σ R) ((algebraMap A R) a) * f := by
+      rw [← smul_eq_C_mul, IsScalarTower.algebraMap_smul]
+    simp only [mul_pow, Finsupp.prod_mul, ← map_pow , ← monomial_one_eq, this]
+    simp only [map_finsupp_prod, map_pow]
+  · intro x _
+    rw [algebra_compatible_smul R, smul_eq_C_mul]
+
+theorem MvPowerSeries.monomial_eq
+    {σ : Type*} {R : Type*} [CommSemiring R]
+    (e : σ →₀ ℕ) (r : σ → R) :
+    MvPowerSeries.monomial R e (e.prod (fun s n => r s ^  n))
+      = (Finsupp.prod e fun s e => (r s • MvPowerSeries.X s) ^ e) := by
+  rw [prod_smul_X_eq_smul_monomial_one, ← map_smul, smul_eq_mul, mul_one]
+
+theorem MvPowerSeries.monomial_smul_const
+    {σ : Type*} {R : Type*} [CommSemiring R]
+    (e : σ →₀ ℕ) (r : R) :
+    MvPowerSeries.monomial R e (r ^ (Finsupp.sum e fun _ n => n))
+      = (Finsupp.prod e fun s e => (r • MvPowerSeries.X s) ^ e) := by
+  rw [prod_smul_X_eq_smul_monomial_one, ← map_smul, smul_eq_mul, mul_one]
+  simp only [Finsupp.sum, Finsupp.prod, Finset.prod_pow_eq_pow_sum]
+
+section DiscreteUniformity
+
+/-- The discrete uniformity -/
+class DiscreteUniformity (α : Type*) [u : UniformSpace α] : Prop where
+  eq_principal_idRel : uniformity α = Filter.principal idRel
+
+/-- The bot uniformity is the discrete uniformity -/
+instance (α : Type*) : @DiscreteUniformity α ⊥ := by
+  apply @DiscreteUniformity.mk α ⊥ rfl
+
+/-- The discrete uniformity induces the discrete topology  -/
+instance (α : Type*) [UniformSpace α] [DiscreteUniformity α] :
+    DiscreteTopology α := by
+  rw [discreteTopology_iff_singleton_mem_nhds]
+  intro a
+  rw [UniformSpace.mem_nhds_iff]
+  simp only [Set.subset_singleton_iff, DiscreteUniformity.eq_principal_idRel]
+  simp only [Filter.mem_principal, idRel_subset]
+  use Set.diagonal α
+  simp only [Set.mem_diagonal_iff, implies_true, true_and]
+  intro x
+  simp only [UniformSpace.ball, Set.mem_preimage, Set.mem_diagonal_iff]
+  exact fun a => a.symm
+
+/-- The discrete uniformity makes a group a `UniformGroup -/
+@[to_additive "The discrete uniformity makes an additive group a `UniformAddGroup`"]
+instance {R : Type*} [Group R] [UniformSpace R] [DiscreteUniformity R] :
+    UniformGroup R where
+  uniformContinuous_div := fun s hs ↦ by
+    simp only [uniformity_prod, DiscreteUniformity.eq_principal_idRel, Filter.comap_principal,
+      Filter.inf_principal, Filter.map_principal, Filter.mem_principal, Set.image_subset_iff]
+    rintro ⟨⟨x, y⟩, z, t⟩
+    simp only [Set.mem_inter_iff, Set.mem_preimage, mem_idRel, and_imp]
+    rintro ⟨rfl⟩ ⟨rfl⟩
+    exact mem_uniformity_of_eq hs rfl
```

Comment:
This should be split in two parts:
- a product of discrete uniform spaces is again discrete (I'd say you want to use [uniformity_prod_eq_comap_prod](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/Basic.html#uniformity_prod_eq_comap_prod))
- any function from a discrete uniform space is uniformly continuous

Reply:
done in the other PR